### PR TITLE
creusot-contracts: Add Iterator impls for [T;N], &mut I

### PR DESCRIPTION
**Pull request by Lysxia**
These came up in examples I've been playing with.

### CODE REVIEW

**Comment by arnaudgolfouse** [comment]
Could you add some test cases for those?

File: creusot-contracts/src/std/iter.rs [hunk 1]
```diff
@@ -211,3 +213,33 @@ extern_spec! {
...
+    #[open]
+    #[predicate(prophetic)]
+    fn produces(self, visited: Seq<Self::Item>, o: Self) -> bool {
+        (*self).produces(visited, *o)
+    }
+
+    #[open]
+    #[predicate(prophetic)]
+    fn completed(&mut self) -> bool {
+        (**self).completed()
```
Comment (jhjourdan) [inline creusot-contracts/src/std/iter.rs:227]:
```suggestion
        (*self).completed() && ^*self == ^^self
```

Comment (jhjourdan) [reply creusot-contracts/src/std/iter.rs:227]: (I think, if I'm not mistaken.)

File: creusot-contracts/src/std/iter/array.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    fn into_iter_pre(self) -> bool {
+        pearlite! { true }
+    }
+
+    #[trusted]
+    #[predicate(prophetic)]
+    fn into_iter_post(self, _res: Self::IntoIter) -> bool {
+        dead
+    }
+}
```
Comment (jhjourdan) [inline creusot-contracts/src/std/iter/array.rs:48]: Could you please give actual specification which state something usable? As is, I don't see what these specifications allow proving. You could take inspiration from iterators of vectors, for example.

Also, I am a bit surprised that these specifications are accepted by Creusot. Usually, we have very limited support for const generics in Creusot.

File: creusot-contracts/src/std/iter/str.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    #[trusted]
+    #[law]
+    #[requires(inv(a))]
+    #[requires(inv(b))]
+    #[requires(inv(c))]
+    #[requires(a.produces(ab, b))]
+    #[requires(b.produces(bc, c))]
+    #[ensures(a.produces(ab.concat(bc), c))]
+    fn produces_trans(a: Self, ab: Seq<Self::Item>, b: Self, bc: Seq<Self::Item>, c: Self) {}
+}
```
Comment (jhjourdan) [inline creusot-contracts/src/std/iter/str.rs:31]: Same here. This specification says nothing.

**Comment by Lysxia** [comment]
I removed `Lines` from this PR because strings are a pain. I added specs for the remaining impls and basic test examples.

**Pull request merged**